File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/filtros/ListaFiltrableAdapter.java
package com.victorlh.android.framework.listas.adapter.filtros;

import com.victorlh.android.framework.listas.ItemLista;
import com.victorlh.android.framework.listas.adapter.AbstractLista;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ListaFiltrableAdapter<T extends ItemLista> implements ListaFiltrable<T> {

	protected AbstractLista<T> adapter;
	protected List<T> listaOriginal;
	protected final Set<FiltroLista<T>> filtros;

	public ListaFiltrableAdapter(AbstractLista<T> adapter) {
		this.filtros = new HashSet<>();
		this.adapter = adapter;
	}

	public void filtrar() {
		if (listaOriginal == null) {
			listaOriginal = adapter.getLista();
		}

		if (filtros.isEmpty()) {
			adapter.setLista(listaOriginal);
			listaOriginal = null;
		} else {
			ArrayList<T> filtrados = new ArrayList<>();
			for (T elemento : listaOriginal) {
				if (isDentroFiltros(elemento)) {
					filtrados.add(elemento);
				}
			}
			adapter.setLista(filtrados);
		}
	}

	@Override
	public void invalidate() {
		this.listaOriginal = null;
	}


	@Override
	public void addFiltro(FiltroLista<T> filtro) {
		this.filtros.add(filtro);
	}

	@Override
	public void removeFiltro(FiltroLista<T> filtro) {
		this.filtros.remove(filtro);
	}

	@Override
	public void removeAllFiltros() {
		this.filtros.clear();
	}

	private boolean isDentroFiltros(T elemento) {
		for (FiltroLista<T> filtro : filtros) {
			boolean dentroFiltro = filtro.isDentroFiltro(elemento);
			if (!dentroFiltro) {
				return false;
			}
		}
		return true;
	}

}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/filtros/ListaFiltrableAdapter.kt
package com.victorlh.android.framework.listas.adapter.filtros

import com.victorlh.android.framework.listas.ItemLista
import com.victorlh.android.framework.listas.adapter.AbstractLista
import java.util.*

class ListaFiltrableAdapter<T : ItemLista?>(protected val adapter: AbstractLista<T>) : ListaFiltrable<T> {

    protected var listaOriginal: List<T>? = null
    protected val filtros: MutableSet<FiltroLista<T>> = HashSet()

    override fun filtrar() {
        if (listaOriginal == null) {
            listaOriginal = adapter.getLista()
        }
        if (filtros.isEmpty()) {
            adapter.setLista(listaOriginal!!)
            listaOriginal = null
        } else {
            val filtrados = ArrayList<T>()
            for (elemento in listaOriginal!!) {
                if (isDentroFiltros(elemento)) {
                    filtrados.add(elemento)
                }
            }
            adapter.setLista(filtrados)
        }
    }

    override fun invalidate() {
        listaOriginal = null
    }

    override fun addFiltro(filtro: FiltroLista<T>) {
        filtros.add(filtro)
    }

    override fun removeFiltro(filtro: FiltroLista<T>) {
        filtros.remove(filtro)
    }

    override fun removeAllFiltros() {
        filtros.clear()
    }

    private fun isDentroFiltros(elemento: T): Boolean {
        for (filtro in filtros) {
            val dentroFiltro = filtro.isDentroFiltro(elemento)
            if (!dentroFiltro) {
                return false
            }
        }
        return true
    }
}